frb_dart/lib/src/ffi/io.dart

```diff
@@ -51,26 +37,137 @@ class DartApiDl {
+      case DartCObjectType.DartNativePointer:
+      case DartCObjectType.DartUnsupported:
+      case DartCObjectType.DartNumberOfTypes:
+      default:
+        throw "Can't read invalid data type $ty";
     }
-    return Uint8List.fromList(ptr.asTypedList(len));
   }
 
-  bool get isSuccess => success > 0;
+  static dynamic _typedDataIntoDart(
+    int ty,
+    ffi.Pointer<ffi.Uint8> typedValues,
+    int nValues, {
+    required bool copy,
+  }) {
+    switch (ty) {
+      case DartTypedDataType.ByteData:
+        final view = typedValues.cast<ffi.Uint8>().asTypedList(nValues);
+        final bytes = copy ? view : Uint8List.fromList(view);
+        return ByteData.view(bytes.buffer);
+      case DartTypedDataType.Int8:
+        final view = typedValues.cast<ffi.Int8>().asTypedList(nValues);
+        return copy ? Int8List.fromList(view) : view;
+      case DartTypedDataType.Uint8:
+        final view = typedValues.cast<ffi.Uint8>().asTypedList(nValues);
+        return copy ? Uint8List.fromList(view) : view;
+      case DartTypedDataType.Int16:
+        final view = typedValues.cast<ffi.Int16>().asTypedList(nValues);
+        return copy ? Int16List.fromList(view) : view;
+      case DartTypedDataType.Uint16:
+        final view = typedValues.cast<ffi.Uint16>().asTypedList(nValues);
+        return copy ? Uint16List.fromList(view) : view;
+      case DartTypedDataType.Int32:
+        final view = typedValues.cast<ffi.Int32>().asTypedList(nValues);
+        return copy ? Int32List.fromList(view) : view;
+      case DartTypedDataType.Uint32:
+        final view = typedValues.cast<ffi.Uint32>().asTypedList(nValues);
+        return copy ? Uint32List.fromList(view) : view;
+      case DartTypedDataType.Int64:
+        final view = typedValues.cast<ffi.Int64>().asTypedList(nValues);
+        return copy ? Int64List.fromList(view) : view;
+      case DartTypedDataType.Uint64:
+        final view = typedValues.cast<ffi.Uint64>().asTypedList(nValues);
+        return copy ? Uint64List.fromList(view) : view;
+      case DartTypedDataType.Float32:
+        final view = typedValues.cast<ffi.Float>().asTypedList(nValues);
+        return copy ? Float32List.fromList(view) : view;
+      case DartTypedDataType.Float64:
+        final view = typedValues.cast<ffi.Double>().asTypedList(nValues);
+        return copy ? Float64List.fromList(view) : view;
+
+      case DartTypedDataType.Uint8Clamped:
+      case DartTypedDataType.Float32x4:
+      case DartTypedDataType.Invalid:
+      default:
+        throw "Can't read invalid typed data type $ty";
+    }
+  }
+
+  static final _externalTypedDataFinalizer =
+      Finalizer<DartNativeExternalTypedData>((externalTypedData) {
+    final handleFinalizer = externalTypedData.callback
+        .cast<ffi.NativeFunction<NativeExternalTypedDataFinalizer>>()
+        .asFunction<DartExternalTypedDataFinalizer>();
+    handleFinalizer(externalTypedData.data, externalTypedData.peer);
+  });
```

Comment:
We need to ensure the finalizer part is really executed (in our valgrind environment), because GC may not happen at all (if memory pressure low).

To do this, maybe have a look at https://github.com/fzyzcjy/flutter_rust_bridge/blob/master/frb_example/pure_dart/dart/lib/main_with_vm_service.dart

Reply:
This one gave me some struggle. I came out with a solution involving some testing code inside the lib itself... Feels dirty but I really didn't find anything better.

Because tests are `dart compile/run`'d, I couldn't check for `Platform` test setting, which means this code won't be stripped down at compile time =/ Even though impact is minimal, any better option is welcome.